src/sinks/mqtt/config.rs

```diff
@@ -0,0 +1,197 @@
+use std::time::Duration;
+
+use rumqttc::{MqttOptions, QoS, TlsConfiguration, Transport};
+use snafu::{ResultExt, Snafu};
+use vector_lib::codecs::JsonSerializerConfig;
+
+use crate::template::Template;
+use crate::{
+    codecs::EncodingConfig,
+    config::{AcknowledgementsConfig, Input, SinkConfig, SinkContext},
+    sinks::{
+        mqtt::sink::{ConfigurationSnafu, MqttConnector, MqttError, MqttSink, TlsSnafu},
+        prelude::*,
+        Healthcheck, VectorSink,
+    },
+    tls::{MaybeTlsSettings, TlsEnableableConfig},
+};
+
+/// Configuration for the `mqtt` sink
+#[configurable_component(sink("mqtt"))]
+#[derive(Clone, Debug)]
+pub struct MqttSinkConfig {
+    /// MQTT server address (The broker’s domain name or IP address).
+    #[configurable(metadata(docs::examples = "mqtt.example.com", docs::examples = "127.0.0.1"))]
+    pub host: String,
+
+    /// TCP port of the MQTT server to connect to.
+    #[serde(default = "default_port")]
+    pub port: u16,
+
+    /// MQTT username.
+    pub user: Option<String>,
+
+    /// MQTT password.
+    pub password: Option<String>,
+
+    /// MQTT client ID.
+    #[serde(default = "default_client_id")]
+    pub client_id: String,
+
+    /// Connection keep-alive interval.
+    #[serde(default = "default_keep_alive")]
+    pub keep_alive: u16,
+
+    /// If set to true, the MQTT session is cleaned on login.
+    #[serde(default = "default_clean_session")]
+    pub clean_session: bool,
+
+    #[configurable(derived)]
+    pub tls: Option<TlsEnableableConfig>,
+
+    /// MQTT publish topic (templates allowed)
+    pub topic: Template,
+
+    #[configurable(derived)]
+    pub encoding: EncodingConfig,
+
+    #[configurable(derived)]
+    #[serde(
+        default,
+        deserialize_with = "crate::serde::bool_or_struct",
+        skip_serializing_if = "crate::serde::skip_serializing_if_default"
+    )]
+    pub acknowledgements: AcknowledgementsConfig,
+
+    #[configurable(derived)]
+    pub quality_of_service: MqttQoS,
+}
+
+/// Supported Quality of Service types for MQTT.
+#[configurable_component]
+#[derive(Clone, Copy, Debug, Derivative)]
+#[derivative(Default)]
+#[serde(rename_all = "lowercase")]
+#[allow(clippy::enum_variant_names)]
+pub enum MqttQoS {
+    /// AtLeastOnce.
+    AtLeastOnce,
+
+    /// AtMostOnce.
+    AtMostOnce,
+
+    /// ExactlyOnce.
+    #[derivative(Default)]
+    ExactlyOnce,
+}
+
+impl From<MqttQoS> for QoS {
+    fn from(value: MqttQoS) -> Self {
+        match value {
+            MqttQoS::AtLeastOnce => QoS::AtLeastOnce,
+            MqttQoS::AtMostOnce => QoS::AtMostOnce,
+            MqttQoS::ExactlyOnce => QoS::ExactlyOnce,
+        }
+    }
+}
+
+const fn default_port() -> u16 {
+    1883
+}
+
+fn default_client_id() -> String {
+    "vector".to_string()
```

Comment:
This can be dangerous since two different clients cannot use the same client ID. So if you have two vector instances that should send mqtt messages to the same sink, this could lead to the clients fighting over the same ID disconnecting each other in an endless loop.

I think there must have been the same issue with other sources or sinks, is there some kind of vector function to create a unique client ID? Ideally, it should be stable between restarts but that's more important for a source than a sink.

Otherwise, I think we should use at least something like `VectorSink<short hash>` and document that multiple mqtt sinks (and sources) connecting to the same broker must not use the same client ID.